### ISSUE TITLE
DH_check() changes in 1.1.1 cause pyca cryptography external test failures

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#9792]
PR #9435 (commit  ddd16c2fe98)  made changes to DH_check() in the 1.1.1 branch. This seems to have caused Travis to fail on the pyca cryptography external tests.

After some digging I've managed to create a standalone C test program that replicates the python test and demonstrates the change in behaviour before and after  ddd16c2fe98.

````
#include <stdio.h>
#include <openssl/bn.h>
#include <openssl/dh.h>

int main(void)
{
    DH *dh = DH_new();
    BIGNUM *p, *g;
    int ret = 1;
    int errflags = 0;

    p = BN_new();
    g = BN_new();
    if (dh == NULL || p == NULL || g == NULL)
        goto err;

    if (!BN_set_word(p, 23) || !BN_set_word(g, 18))
        goto err;

    if (!DH_set0_pqg(dh, p, NULL, g))
        goto err;
    p = g = NULL;

    if (!DH_check(dh, &errflags))
        goto err;

    /* We expect errflags to be non-zero */
    if (errflags == 0)
        goto err;

    ret = 0;

 err:
    BN_free(p);
    BN_free(g);
    DH_free(dh);

    if (ret == 0)
        printf("Success\n");
    else
        printf("Failure\n");
    return ret;
}
````
Prior to the change this test program will print "Success", and after it, it prints "Failure".

It seems that before this change the errflags were being set to `DH_UNABLE_TO_CHECK_GENERATOR`, but are now not set to anything. The PR in question seems to have removed the generator checks completely. Why is this?

Possibly the pyca test needs to change because it assumes something about the internal implementation of DH_check() which is no longer true (i.e. that it is unable to check certain values of generator).

**Comment by mattcaswell** [comment]
@bernd-edlinger @reaperhulk

**Comment by bernd-edlinger** [comment]
Ah, well, is is unfortunate that we had all the time test failures so I really
had no chance to spot this eariler.

Yes, previously generators were
- g=2 and p=11 mod 24
- g=5 and p=3 or 5 mod 10

so what is exactly not state of the art i.e. order 2q.
since we change that to generate
- g=2 and p=23 mod 24
- g=3 and p=11 mod 12
- g=5 and p=59 mod 60

we need to relax the rules what DH_check accepts
so currently any p that is safe prime, and g != 0, 1, p-1,
in order to interoperable with DH params generated by 3.0.0

Previously p=23, g=18 was rejected. now it is accepted by DH_check.
However I am not sure if we should check that p is min. 512 bits.
master we have now a lower limit of 512 bits for p.
But that is only checked in dh_builtin_genparams,
dh_generate_key and dh_compute_key.
However I forgot about adding that check in DH_check, or
better DH_check_params.

**Comment by bernd-edlinger** [comment]
one thing is wrong in the test case:
when DH_check did return 1, we expect errflags to be zero:
```C
    if (!DH_check(dh, &errflags))
        goto err;

    /* We expect errflags to be zero */
    if (errflags == 0)
        goto err;
```

that would pass both before and after that change.

**Comment by bernd-edlinger** [comment]
Err, actually this would be invariant:
```C
    if (!DH_check(dh, &errflags)) {
        /* We expect errflags to be non-zero */
        if (errflags == 0)
            goto err;
    } else {
        /* We expect errflags to be zero */
        if (errflags != 0)
            goto err;
    }
```

**Comment by bernd-edlinger** [comment]
Hmm, #9796 adds check for allowed key-length to DH_check but it does not
seem to fix the pyca tests.

**Comment by bernd-edlinger** [comment]
>       assert backend.dh_parameters_supported(23, 5)
>       assert not backend.dh_parameters_supported(23, 18)

for now I would suggest to change that to
`assert not backend.dh_parameters_supported(23, 22)`
for instance, since that creates a order 2 subgroup.
But I think we will raise the minimum bit length from 512 to 1024 soon.
So I would suggest using larger primes in the good test to be safe.

**Comment by mattcaswell** [comment]
> one thing is wrong in the test case:
when DH_check did return 1, we expect errflags to be zero:

That's not correct. errflags can be non-zero even if DH_check returns 1.

**Comment by bernd-edlinger** [comment]
You are right, I mixed that up...
if it returns zero, something like out-of-memory may have happened, and then
errflags is still zero.

**Comment by reaperhulk** [comment]
Can this result in situations where DH parameters generated by 1.1.1d would be rejected by 1.1.1c and earlier? If so, that's potentially a problem since users would generally assume a bugfix release wouldn't cause compatibility issues like that. `pyca/cryptography` can easily change the test though, so if there's no compat concern we'll just go ahead and make the change. Let us know :)

**Comment by bernd-edlinger** [comment]
Yes, that could be possible.
While I think the change in DH_check need to stay (and DH_check is not really used often),
we could partially revert the change in dh_builtin_genparams for 1.1.1
I would not completely revert, since the change did also improve the performance quite
a lot.

**Comment by bernd-edlinger** [comment]
See #9820

**Comment by bernd-edlinger** [comment]
> we'll just go ahead and make the change. Let us know :)

yes, please change the bad test to use something really bad, like prime, but not safe prime,
and please use 1024 bit primes, to be safe.

**Comment by mattcaswell** [comment]
This seems to have been fixed up stream. We need to sync our external test submodule to the latest version.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input
